sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala

```diff
@@ -153,6 +151,27 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
 
       // --- BroadcastHashJoin --------------------------------------------------------------------
 
+      case ExtractEquiJoinKeys(joinType, leftKeys, rightKeys, condition, left, right)
+        if canBuildRight(joinType) && canBuildLeft(joinType)
+          && left.stats.hints.broadcast && right.stats.hints.broadcast =>
```

Comment:
```Scala
if canBuildRight(joinType) && canBuildLeft(joinType)
  && (left.stats.hints.broadcast || right.stats.hints.broadcast) =>
```

Can you combine these three cases?

Reply:
I think we can create new methods for it
```
def shouldBuildLeft(joinType: JoinType, left: LogicalPlan, right: LogicalPlan): Boolean {
  if (left.stats.hints.broadcast) {
    if (canBuildRight(joinType) && right.stats.hints.broadcast) {
      // if both sides have broadcast hint, only broadcast left side if its estimated pyhsical size is smaller than right side
      left.stats.sizeInBytes <= right.stats.sizeInBytes
    } else {
      // if only left side has the broadcast hint, broadcast the left side.
      true
    }
  } else {
      if (canBuildRight(joinType) && right.stats.hints.broadcast) {
        // if only right side has the broadcast hint, do not broadcast the left side.
        false
      } else {
        // if neither one of the sides has broadcast hint, only broadcast the left side if its estimated physical size is smaller than the treshold and smaller than right side.
        canBroadcast(left) && left.stats.sizeInBytes <= right.stats.sizeInBytes
      }
  }
}

def shouldBuildRight...
```
and use it like
```
case ExtractEquiJoinKeys(joinType, leftKeys, rightKeys, condition, left, right)
    if canBuildRight(joinType) && shouldBuildRight(joinType, left, right)
```

Reply:
I have an idea to combine these three cases to `broadcastBuildSide` function:
https://github.com/apache/spark/blob/a2cf0ced6615e342e84e270cf060ace69eb3ba0f/sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala#L150-L185